src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -515,10 +515,26 @@ private static MethodHandle makeSimpleConcat(MethodType mt, String[] constants)
             mh = simpleConcat3(paramType0);
             mh = MethodHandles.insertArguments(mh, 2, suffix);
             return MethodHandles.insertArguments(mh, 0, prefix);
-        } else if (paramCount == 2 && !mt.hasPrimitives() && suffix.isEmpty()
-                && constants[0].isEmpty() && constants[1].isEmpty()) {
+        } else if (paramCount == 2 && constants[1].isEmpty()) {
+            var paramType1 = mt.parameterType(1);
             // Two reference arguments, no surrounding constants
-            return simpleConcat();
+            if (!mt.hasPrimitives() && suffix.isEmpty() && prefix.isEmpty()) {
+                return simpleConcat();
+            }
+
+            if (suffix.isEmpty() && paramType1 == String.class) {
+                // "prefix" + x + s
+                mh = simpleConcat3(paramType0);
+                mh = MethodHandles.insertArguments(mh, 0, prefix);
+                return MethodHandles.filterArguments(mh, 1, objectStringifier());
```

Comment:
While this is a fun trick, but seems like there's a non-trivial cost here? We'd go down different paths and generate different classes for `"foo" + bar + baz` and `"foo" + bar + " .. " + baz` with this. Special casing when we get the added shapes for more or less free (plain `simpleConcat()`) is a different matter but here you need to construct a new couple of shapes with `insert-` and `filterArguments`.

(Check on paramType1 could be `!paramType1.isPrimitive()`)

Reply:
I'm also not sure how much cost the simpleConcat handling of the 2 parameters would bring, I've removed that, and this is more appropriately implemented by the InlineHiddenClassStrategy.

Reply:
If we handle the null values of prefix and suffix, the cost of supporting the 2-parameter scenario will be much lower.